shtrom's plugins cleanup

shtrom opened the pull request:
Cleanup some of my plugins (network/upnpc_, isp/internode) a bit.
* Stricter shell syntax
* Add support for host_name option  to allow displaying graph for another host (e.g., the uplink router)
* Add multigraph support for upnpc (backwards compatible)
* Deprecate upnpc_bitrate (backward compatible), and optionally show on upnpc_traffic instead

shtrom commented:
Just found a bug in `ssl/ssl-certificate-expiry`, sneaking the fix here, too.

sumpfralle commented on plugins/isp/internode_usage:
```diff
@@ -68,46 +70,74 @@ fi
-	xpath -q -n -e "${xpath}" | \
-		sed 's/.*="\([^"]\+\)".*/\1/'
+	# shellcheck disable=SC2064
+	trap "error extracting attribute at ${xpath} >&2" EXIT
```
How about using the `curl ... || echo >&2 "..."` approach instead?
(I think, it is easier to read here)

Another detail: I think, you omitted `echo` in the `trap`.

shtrom replied:
Good catch on the missing `echo`.

I can't remember why I didn't just do `cmd || echo >&2`. I'll need to rethink that 😬

shtrom replied:
Ah! Yes! That's it, you're right. I want to be able to output the error message, but to let the caller decide whether this is an error that can be tolerated (with `|| true`, which is used a few times).

So, yeah, I think we need to keep it like that.

sumpfralle replied:
> Ah! Yes! That's it, you're right. [..]

A good example of code with surprising side effects :)

I would appreciate a more obvious approach, but I could live with this one, too.

shtrom replied:
Ah! I think I found a better way. What about now?

sumpfralle commented on plugins/network/upnpc_:
```diff
@@ -37,21 +39,38 @@ SPDX-License-Identifier: GPL-2.0-or-later
+    set -x
+fi
+
+if ! command -v upnpc >/dev/null; then
```
This test should be part of the `autoconf` handling.
At the moment a call for `autoconf` results in an invalid response string (`upnpc not found ...`) instead of `no (upnpc not found)`.

Probably you should move the data retrieval (`DATA=$(upnpc -s)`) down to the `config` and `fetch` operation in order to avoid the complexities auf combining `autoconf` and central data collection?

sumpfralle commented on plugins/isp/internode_usage:
```diff
@@ -221,14 +221,14 @@ graph_data() {
 		*)
-			echo "usage.value ${SERVICE_USAGE:-U}"
-			echo "ideal.value ${IDEAL_USAGE:-U}"
+			echo "usage.value ${CURRENT_TIMESTAMP}:${SERVICE_USAGE:-U}"
```
Is it really necessary to add the timestamp here? This sounds like a bug in the spool processing to me.

shtrom replied:
It is, but there's not real way around it \:

sumpfralle replied:
If it is a bug: where is the ticket? (I would like to solve it instead of fostering bad examples in plugin code).

shtrom replied:
Actually, maybe it's not a bug as such, but a corner case.

This particular plugin outputs data with two different dates: right now, and more or less a day before. Munin is not used to it, and assumes that all the data from a plugin comes from the same time, which is not the case here. Not stating the timestamp explicitely here leads munin to assume that it's the same as the delayed one, which is often lower than the last received statement, which leads Munin to discard fresh values from the current time.

I think it's more of a weird interaction between Munin's expectations, and the data from the API this plugin queries.

sumpfralle commented on plugins/network/upnpc_:
```diff
@@ -20,7 +20,11 @@ Linux systems with upnpc installed (miniupnpc package).
+If you do not want to show the link maximum bitrates, add the following
+plugin-configuration:
+
+  [upnpc*]
```
Maybe add the final underscore in order to keep it focused?

shtrom replied:
No, as the plugin also supports `upnpc` as a shorthand for `upnpc_multi`

sumpfralle replied:
Good point!

sumpfralle commented on plugins/network/upnpc_:
```diff
@@ -144,6 +147,14 @@ down.graph no
 up.negative down
 EOF
 			;;
+		"multi"|"upnpc")
```
I think, it would be better to normalize the "MODE" value at the beginning (i.e. `[ "$MODE" = "upnpc" ] && MODE="multi"`). Otherwise the `"multi"|"upnpc"` conditional can be easily misunderstood.

sumpfralle commented on plugins/network/upnpc_:
```diff
@@ -144,6 +147,14 @@ down.graph no
 			;;
+		"multi"|"upnpc")
+			echo "multigraph ${PLUGIN_NAME}"
+			config "traffic"
```
Is the `traffic` really suitable as a *top level summary* of the other graphs?
Otherwise you may just want to graph all of them at the same level? (just replace the dot with an underscore)

shtrom replied:
I think it is. It's a matter of preference though.

That said, we need at one top-level graph. We can't just have subpages.

sumpfralle replied:
> That said, we need at one top-level graph. We can't just have subpages.

I wanted to propose to use underscores instead of dots in the multigraph names. In this case there would be no hierarchy and thus just a flat list of graphs at the top level.

Feel free to pick the most suitable approach for this set of information.

shtrom replied:
Ah right, yeah. I never thought about it, actually.

That said, one of the reason I (ab)use multigraphs is to not have too many graphs on the front page, while still being able to drill down to more details if I want to. In this respect, I find the traffic most useful as a top-level graph, while the others are less so.

I think I'm going to leave it as is (:

sumpfralle commented:
Thanks a lot for this good amount of change proposals!

I commented a few of them in detail ...

shtrom commented:
@sumpfralle Ready for another look. I've fixed what I thought was fixable, and left comments explaining why I didn't think the rest was.

Let me know what you think.

shtrom commented:
I just snuck in an additional fix for the fresh-backups plugins, too, as the limits where calculated incorrectly.

sumpfralle commented:
> Let me know what you think.

Thank you! I did not notice any new details, that I would like to discuss.

> I just snuck in an additional fix for the fresh-backups plugin, too, as the limits were calculated incorrectly.

In general it is fine. In a perfect world I would prefer, if you could separate "rename" and "change" into two commits. At the moment the change within the renamed plugin is hard to spot.

What do you think: are there still things, you would like to improve before I merge the commits?

shtrom commented:
Couple more fixups. I split the move/change commit from `fresh-backup`, and added caching to the `internode_usage` plugin, as their API is a bit flakey.

I think I also found a `trap`-less way to report errors while letting the caller decide what to do about it.

sumpfralle commented on plugins/isp/internode_usage:
```diff
@@ -104,6 +104,25 @@ fetch() {
 
+get_cached_api_xml() {
+	API_XML=''
+	API_XML_CACHE="${MUNIN_PLUGSTATE}/$(basename "${0}").cache"
```
I would suggest to move this up closer to the top of the plugin code. This will make it more visible, that cached data is used.

shtrom replied:
Hum, I prefer the code organised like this, with helper functions together, and data-collecting functions together.

However, I added some details about the caching mechanism in the documentation.

sumpfralle replied:
I only meant the definition of the variable (`API_XML_CACHE`), as this filename is part of the interface of the plugin. But anyway: personal preferences vary (which is good) ... :)

shtrom replied:
Oh right, sorry. Yeah, it was only one line up, but I see your point. It will happen next time I have improvements to add!

sumpfralle commented on plugins/isp/internode_usage:
```diff
@@ -104,6 +104,25 @@ fetch() {
 }
 
+get_cached_api_xml() {
+	API_XML=''
```
Probably this variable (especially with such a generic name) deserves to be `local`?

sumpfralle commented on plugins/isp/internode_usage:
```diff
@@ -83,31 +83,25 @@ fi
-		| sed 's/.*="\([^"]\+\)".*/\1/'
-	trap - EXIT
+	(xpath -q -n -e "${xpath}" \
+		| sed 's/.*="\([^"]\+\)".*/\1/') \
```
I think, here (and above in the other function) you merely test, wether `sed` failed before.

I guess, you will need to store the output of `xpath` in a variable :)

shtrom replied:
Ah, you're right...

This is annoying, as I was trying to keep it short and sweet \:

sumpfralle commented on plugins/isp/internode_usage:
```diff
@@ -83,31 +83,25 @@ fi
-	trap - EXIT
+	# shellcheck disable=SC2154
+	curl -u "${internode_api_login}:${internode_api_password}" -f ${CURL_ARGS} "$@" \
+		|| (echo "error fetching ${*} for user ${internode_api_login}" >&2 \
```
Just a minor nitpick: `{ ...; }` (instead of `(...)`) would be more efficient here, since it does not require a fork.

shtrom replied:
Heh, TIL. I never realised I could use `{ ... ; }`.

The difference is notable
```
$ time sh -c 'for i in `seq 10000`; do true && (false); done'
sh -c 'for i in `seq 10000`; do true && (false); done'  2.66s user 2.47s system 104% cpu 4.886 total
$ time sh -c 'for i in `seq 10000`; do true && { false; }; done'
sh -c 'for i in `seq 10000`; do true && { false; }; done'  0.02s user 0.00s system 98% cpu 0.019 total
```

sumpfralle replied:
Nice benchmark!

But to be fair: executing non-builtin statements (e.g. calling `curl`) is quite common in a script. Thus the difference in reality is probably not as noticeable as the above numbers may imply.

sumpfralle commented:
It looks good to me!

Feel free to respond to my last comments, if they feel useful for you.

As soon as you rebase your `fixup` commits, I will merge your set of changes.
Thank you!

shtrom commented:
Ok, fixed and squashed.

sumpfralle commented:
Thank you for contributing this great amount of neat improvements!
